### PR TITLE
[CTCP-2922] added business validation to match messageType with the root node

### PR DESCRIPTION
**Pull request by prashanthmamidi**

### CODE REVIEW

File: app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala [hunk 1]
```diff
@@ -95,23 +96,40 @@ class JsonValidationServiceImpl @Inject() extends JsonValidationService {
...
+          Set(CustomValidationMessage(None, "Root node doesn't match with the messageType", isBusinessValidation = true))
+        } else {
+          schemaValidator
+            .validate(jsonNode)
+            .asScala
+            .map(
+              vm => CustomValidationMessage(Some(vm.getSchemaPath), vm.getMessage, isBusinessValidation = false)
+            )
+            .toSet
+        }
```
Comment (drnaylor-oc) [inline app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala:131]: The ticket explicitly said business validation should be in its own service.

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala [hunk 2]
```diff
@@ -76,38 +78,66 @@ class XmlValidationServiceImpl @Inject() (implicit ec: ExecutionContext) extends
...
+                    Either.left(BusinessValidationError("Root node doesn't match with the messageType"))
+                  } else {
+                    NonEmptyList
+                      .fromList(errorBuffer.toList)
+                      .map(
+                        x => Either.left(XmlFailedValidation(x))
+                      )
+                      .getOrElse(parseXml)
+                  }
+
```
Comment (drnaylor-oc) [inline app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:140]: As above.

**Comment by drnaylor-oc** [comment]
We're mixing the schema validation and the business validation here, and that's just going to turn into a nightmare for maintainability. The ticket said that the business validation needed to be in its own service for the reason that we will need to add more business validation rules to the validator -- and instead we've ended up with what's very quickly going to end up with a mix of schema validation code and business validation code.

If we can't do it via it's own service, there still needs to be some clean separation between what is schema validation and business validation. I'm not asking for really clever code that lets us specify the XML and Json validation rules once, but at least something that doesn't just end up with a set of if statements all together in the schema validation method.

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala [hunk 3]
```diff
@@ -76,38 +78,66 @@ class XmlValidationServiceImpl @Inject() (implicit ec: ExecutionContext) extends
...
+                    new DefaultHandler {
+                      var inMessageTypeElement = false
+
+                      override def characters(ch: Array[Char], start: Int, length: Int): Unit =
+                        if (inMessageTypeElement) {
+                          elementValue = new String(ch, start, length)
+                        }
+
+                      override def startElement(uri: String, localName: String, qName: String, attributes: Attributes) = {
+                        if (qName.startsWith("ncts:")) {
```
Comment (drnaylor-oc) [inline app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:98]: You cannot assume that the prefix will be this. Traders can use whatever prefix they like, as long as the prefix and namespace match up.

An assumption like this actually bit me when I was setting up connectivity on QA, ERMIS didn't use the prefixes the EIS spec gave. Still totally valid, but the hardcoded prefixes broke things.

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala [hunk 4]
```diff
@@ -76,38 +78,66 @@ class XmlValidationServiceImpl @Inject() (implicit ec: ExecutionContext) extends
...
+                  var rootTag      = ""
+
+                  val parser = saxParser.newSAXParser.parse(
+                    inputSource,
+                    new DefaultHandler {
+                      var inMessageTypeElement = false
+
+                      override def characters(ch: Array[Char], start: Int, length: Int): Unit =
+                        if (inMessageTypeElement) {
+                          elementValue = new String(ch, start, length)
```
Comment (drnaylor-oc) [inline app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:94]: What do I get if I put the following XML fragment?

```xml
<messageType>
    CC015C
</messageType>
```

specifically with the whitespace?

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala (hunk 3 above)
Comment (drnaylor-oc) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:98]: (that does remind me, I should check this assumption for trader -> ERMIS messages in the router)

Comment (prashanthmamidi) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala]: yes, updated to override startPrefixMapping to get the namespace

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala (hunk 4 above)
Comment (prashanthmamidi) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:157]: if we spaces the validation works

File: app/uk/gov/hmrc/transitmovementsvalidator/models/errors/ErrorCode.scala [hunk 5]
```diff
@@ -27,11 +27,12 @@ import play.api.libs.json.Writes
...
-  val NotFound: ErrorCode             = ErrorCode("NOT_FOUND", NOT_FOUND)
-  val InternalServerError: ErrorCode  = ErrorCode("INTERNAL_SERVER_ERROR", INTERNAL_SERVER_ERROR)
-  val SchemaValidation: ErrorCode     = ErrorCode("SCHEMA_VALIDATION", OK)
-  val UnsupportedMediaType: ErrorCode = ErrorCode("UNSUPPORTED_MEDIA_TYPE", UNSUPPORTED_MEDIA_TYPE)
+  val BadRequest: ErrorCode              = ErrorCode("BAD_REQUEST", BAD_REQUEST)
+  val NotFound: ErrorCode                = ErrorCode("NOT_FOUND", NOT_FOUND)
+  val InternalServerError: ErrorCode     = ErrorCode("INTERNAL_SERVER_ERROR", INTERNAL_SERVER_ERROR)
+  val SchemaValidation: ErrorCode        = ErrorCode("SCHEMA_VALIDATION", OK)
+  val UnsupportedMediaType: ErrorCode    = ErrorCode("UNSUPPORTED_MEDIA_TYPE", UNSUPPORTED_MEDIA_TYPE)
+  val BusinessValidationError: ErrorCode = ErrorCode("BUSINESS_VALIDATION_ERROR", BAD_REQUEST)
```
Comment (drnaylor-oc) [inline app/uk/gov/hmrc/transitmovementsvalidator/models/errors/ErrorCode.scala:35]: This should also be an `OK`, as the input was totally fine, it's just not been validated - same as the schema validation.

I should check to see what `common-transit-convention-traders` expects, now I think about it...

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala (hunk 4 above)
Comment (drnaylor-oc) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:157]: Awesome, thanks for checking!

File: app/uk/gov/hmrc/transitmovementsvalidator/models/errors/ErrorCode.scala (hunk 5 above)
Comment (prashanthmamidi) [reply app/uk/gov/hmrc/transitmovementsvalidator/models/errors/ErrorCode.scala:35]: currently, `common-transit-convention-traders` will pass the  same 400 back to the trader.  As per the ticket we should return 400

Comment (drnaylor-oc) [reply app/uk/gov/hmrc/transitmovementsvalidator/models/errors/ErrorCode.scala:35]: I raised a ticket for this. We'll not worry about it here.

File: conf/json/cc009c-schema.json [hunk 6]
```diff
@@ -14589,9 +14589,6 @@
     },
     "n1:MessageTypes": {
       "description": "The list of codes used to identify the type of messages exchanged on the Common Domain and on the External Domain (e.g. CD001B, CD302A, CC515C, CD801A).The values of this codelist are also available and updated in CS/RD2.",
-      "enum": [
-        "CC009C"
-      ],
```
Comment (drnaylor-oc) [inline conf/json/cc009c-schema.json:14594]: If we're going to do this, we should at least add a pattern in its place to make it much closer to 

```
"pattern": "^C[CD]\\d{3}C$"
```

or make the enum hold all valid values as per the XSD.

Comment (prashanthmamidi) [reply conf/json/cc009c-schema.json]: to be consistent with xsd - added all valid values to the list

**Review by drnaylor-oc — COMMENTED** [review]
I admit, I was expecting an Alpakka based solution rather than reusing the SAX parser and Json reader (like we've used elsewhere in the codebase) because that way we could have just had each element report their errors rather than store state and such (and I would have been able to do something to make it read the stream once rather than twice), but I don't really want this knocking around much longer.

We should make sure the business errors that are returned are consistent from both though.

File: app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala [hunk 7]
```diff
@@ -107,13 +109,45 @@ class JsonValidationServiceImpl @Inject() extends JsonValidationService {
...
 
+  def validateJson(source: Source[ByteString, _])(implicit materializer: Materializer): Try[Set[CustomValidationMessage]] =
+    Using(source.runWith(StreamConverters.asInputStream(20.seconds))) {
+      jsonInput =>
+        val jsonNode: JsonNode      = mapper.readTree(jsonInput)
+        val rootNode                = jsonNode.fields().next().getKey
+        val messageType             = jsonNode.path(rootNode).path("messageType").textValue()
+        val messageTypeFromRootNode = rootNode.split(":")(1)
+        if (!messageTypeFromRootNode.equalsIgnoreCase(messageType)) {
+          Set(CustomValidationMessage(None, "Root node doesn't match with the messageType", isBusinessValidation = true))
```
[the remark this answers is not in the document]
Comment (prashanthmamidi) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala:145]: This `CustomValidationMessage` will translate to  `BuisnessValidationError` later,  this  was originally added to handle both schema and business validations - as we separated this now - will update the code to directly use `BuisnessValidationError ` and delete `CustomValidationMessage`

**Comment by prashanthmamidi** [comment]
> I admit, I was expecting an Alpakka based solution rather than reusing the SAX parser and Json reader (like we've used elsewhere in the codebase) because that way we could have just had each element report their errors rather than store state and such (and I would have been able to do something to make it read the stream once rather than twice), but I don't really want this knocking around much longer.
> 
> We should make sure the business errors that are returned are consistent from both though.

Yes, business errors returned are consistent for both json and xml from validator service.

**Review by drnaylor-oc — APPROVED** [review]
I'll likely revisit this code to make business validation use Alpakka, but that's another ticket. This looks like it works and @SandeepSinghVerma1992 has validated it -- so let's get this in now and hope we can get it on environments soon.

**Pull request merged**